### PR TITLE
Introduce a non-generic Image base class

### PR DESCRIPTION
**Pull request by antonfirsov**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
- Implements the proposal demonstrated in #897.
- Change majority of the extension methods working on `Image<TPixel>` to work on `Image` instead:
```C#
// The non-generic Image.Load overloads no longer return Image<Rgba32>
using (Image image = Image.Load("foo.jpg"))
{
    image.Mutate(x => x.Resize(100, 100);
    image.SaveAsPng(someStream);
}
```

- Touches #621 by making most of the non-generic processors public. (Exception: Histogram processors, because I'm not entirely happy with all namings.)
- Fix a few doc issues while doing the refactor.
- While implementing better test coverage for `Image.Load` variants, discovered and added a few missing overloads

Hope I haven't missed any of the important processors and / or extensions!

To enable extensions using color values as parameters on `Image`, we need to introduce a pixel-agnostic color type. I have [a proposal](https://gist.github.com/antonfirsov/ba3b3fc05e7aba19458d6d61c8b38e2c) for this, planning to PR it within the next week.

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
I'm very happy for the most part with this thus far. I'd like to know more about your concerns with the Histogram naming and would like to help there.

**Comment by antonfirsov** [comment]
`AdaptiveHistEqualizationSWProcessor` - too many abbrevations.

On the other hand, `AdaptiveHistogramEqualizationSlidingWindowProcessor` would be way too long. I don't know what is the good solution.

**Comment by JimBobSquarePants** [comment]
I think we’ve still got a long way to go before we hit the dizzy heights of https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/BoundTree/BoundTreeRewriter.cs#L100

So I’m ok with the longer form

**Comment by antonfirsov** [comment]
Ok, went with the longer names. All processors without color parameters should be `public (sealed)` now.

File: src/ImageSharp/Image{TPixel}.cs [hunk 1]
```diff
@@ -3,26 +3,24 @@
...
 namespace SixLabors.ImageSharp
 {
     /// <summary>
     /// Encapsulates an image, which consists of the pixel data for a graphics image and its attributes.
+    /// For generic <see cref="Image{TPixel}"/>-s the pixel type is known at compile time.
     /// </summary>
     /// <typeparam name="TPixel">The pixel format.</typeparam>
-    public sealed class Image<TPixel> : IImage, IConfigurable
+    public sealed class Image<TPixel> : Image
         where TPixel : struct, IPixel<TPixel>
```
Comment (iamcarbon) [inline src/ImageSharp/Image{TPixel}.cs]: Consider unmanaged constraint?

**Comment by antonfirsov** [comment]
I'm merging this to go on with the rest of #907. If I made a mistake, I will be able to fix it in the follow up PR-s.

**Pull request merged**